### PR TITLE
SPDY parser.js tests

### PR DESCRIPTION
**Pull request by diasdavid**
Hi @indutny putting this here as a WIP. I (finally) got the test case working, however noticed that for a SYN_FRAME (type 0x01), the callback was being called with type being "HEADERS".

Also, found this - https://github.com/indutny/spdy-transport/blob/master/lib/spdy-transport/protocol/spdy/parser.js#L184 which I guess to be leftovers from http2, at least I can't find any reference on the SPDY spec about a PUSH_PROMISE type of frame. Same goes for PRIORITY https://github.com/indutny/spdy-transport/blob/master/lib/spdy-transport/protocol/spdy/parser.js#L191

Will continue working on the other frame tests, just wanted to keep you posted :)

### CODE REVIEW

File: lib/spdy-transport/protocol/spdy/parser.js [hunk 1]
```diff
@@ -172,7 +173,7 @@ Parser.prototype.onSynHeadFrame = function onSynHeadFrame(type,
 
     if (!isPush) {
       callback(null, {
-        type: 'HEADERS',
+        type: 'SYN_STREAM',
```
Comment (indutny) [inline lib/spdy-transport/protocol/spdy/parser.js]: This is going to break everything! ;) I did it for compatibility with HTTP2.

File: lib/spdy-transport/protocol/spdy/parser.js [hunk 2]
```diff
@@ -145,6 +145,7 @@ Parser.prototype.onSynHeadFrame = function onSynHeadFrame(type,
 
   var head = body.clone(offset);
   body.skip(offset);
+
```
Comment (indutny) [inline lib/spdy-transport/protocol/spdy/parser.js]: Superfluous change.

**Comment by indutny** [comment]
Yay!

File: lib/spdy-transport/protocol/spdy/parser.js (hunk 1 above)
Comment (diasdavid) [inline lib/spdy-transport/protocol/spdy/parser.js]: So, am I interfering with your idea/plan? Feel free to enlighten me :)

Comment (indutny) [inline lib/spdy-transport/protocol/spdy/parser.js]: I'd go with just HEADERS for both SYN_STREAM and SYN_REPLY. HTTP2 is using just HEADERS for both of them, and for continuation headers too. There is no complexity in handling this in `connection.js`.

TL;DR - let's leave it as it is ;)

**Comment by diasdavid** [comment]
What about the PUSH_PROMISE and PRIORITY?

**Comment by indutny** [comment]
@diasdavid what about it?

File: lib/spdy-transport/protocol/spdy/parser.js (hunk 1 above)
Comment (diasdavid) [inline lib/spdy-transport/protocol/spdy/parser.js]: Got it, does that mean that it sends back a HEADERS type code instead of the SYN REPLY? Didn't notice that, or does this just happens on the parsing part.

**Comment by diasdavid** [comment]
Is it a compatibility thing too? (The fact they are around on the spdy/parser.js)

File: lib/spdy-transport/protocol/spdy/parser.js (hunk 1 above)
Comment (indutny) [inline lib/spdy-transport/protocol/spdy/parser.js]: The framer has them apart: `requestFrame`, `responseFrame`, `headersFrame`. The particular implementation chooses the right frame type.

**Comment by indutny** [comment]
Yeah, exactly. The protocols should look completely the same to the `connection` and `stream`.

**Comment by indutny** [comment]
Hello @diasdavid ! Do you need any help with this?

File: test/spdy/parser-test.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        priority: {
+          exclusive: false,
+          parent: 0,
+          weight: 1.
+        },
+        type: 'HEADERS' // by spec 'SYN_STREAM'
+      }, done);
+    })
+
+    it('should parse frame with http header and Unidirectional flag', function(done) {
```
Comment (diasdavid) [inline test/spdy/parser-test.js]: When `0x02 = FLAG_UNIDIRECTIONAL` is passed, nothing is declared on the frame parsed. Is this intended @indutny ?

File: test/spdy/parser-test.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+          exclusive: false,
+          parent: 0,
+          weight: 1.
+        },
+        type: 'HEADERS' // by spec 'SYN_STREAM'
+      }, done);
+    })
+
+    /*
+    it('should fail on stream ID 0', function(done) {
```
Comment (diasdavid) [inline test/spdy/parser-test.js]: the parser doesn't fail with stream Id 0. Probably this validation is done in another place, could you confirm @indutny ?

**Comment by diasdavid** [comment]
@indutny sorry, got consumed on another things, doing the rest of the tests now :)

File: test/spdy/parser-test.js (hunk 3 above)
Comment (indutny) [inline test/spdy/parser-test.js]: Just fixed it, thanks for reminding!

File: test/spdy/parser-test.js (hunk 4 above)
Comment (indutny) [inline test/spdy/parser-test.js]: Fixed too, thank you!

**Comment by diasdavid** [comment]
Any feedback on things that could be done to make this better? :) All tests pass now, code style is followed. Frames were extracted from go<->node experiments (which all of them worked :) )

File: lib/spdy-transport/protocol/spdy/parser.js [hunk 5]
```diff
@@ -190,7 +190,7 @@ Parser.prototype.onSynHeadFrame = function onSynHeadFrame(type,
 
     if (!isPush) {
       callback(null, {
-        type: 'HEADERS',
+        type: 'HEADERS', // type === 0x01 ? 'SYN_STREAM': 'SYN_REPLY',
```
Comment (indutny) [inline lib/spdy-transport/protocol/spdy/parser.js]: Superfluous change.

File: test/spdy/parser-test.js [hunk 6]
```diff
@@ -0,0 +1,445 @@
+var assert = require('assert');
+
+var transport = require('../../');
+var spdy = transport.protocol.spdy;
+
+describe('SPDY Parser', function() {
```
Comment (indutny) [inline test/spdy/parser-test.js]: I think this is SPDYv3-only test, let's rename the file to `v3-parser-test.js`.

File: test/spdy/parser-test.js [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+  var parser;
+
+  beforeEach(function() {
+    var pool = spdy.compressionPool.create();
+    parser = spdy.parser.create({});
+    var comp = pool.get('3.1');
+    parser.setCompression(comp);
```
Comment (indutny) [inline test/spdy/parser-test.js]: I think it might be a good idea to skip preface here instead of duplicating the logic in `pass`/`fail`.

File: test/spdy/parser-test.js [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+        path: '/',
+        priority: {
+          exclusive: false,
+          parent: 0,
+          weight: 1.
+        },
+        type: 'HEADERS', // by spec 'SYN_STREAM'
+        writable: false
+      }, done);
+    })
```
Comment (indutny) [inline test/spdy/parser-test.js]: Missing semicolon

File: test/spdy/parser-test.js [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+        path: '/',
+        priority: {
+          exclusive: false,
+          parent: 0,
+          weight: 1.
+        },
+        type: 'HEADERS', // by spec 'SYN_STREAM'
+        writable: true
+      }, done);
+    })
```
Comment (indutny) [inline test/spdy/parser-test.js]: Ditto.

File: test/spdy/parser-test.js [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
```
Comment (indutny) [inline test/spdy/parser-test.js]: ditto.

File: test/spdy/parser-test.js [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+        priority: {
+          exclusive: false,
+          parent: 0,
+          weight: 1.
+        },
+        type: 'HEADERS', // by spec 'SYN_STREAM'
+        writable: false
+      }, done);
+    })
+
```
Comment (indutny) [inline test/spdy/parser-test.js]: Superfluous newline.

File: test/spdy/parser-test.js [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+        path: undefined,
+        priority: {
+          exclusive: false,
+          parent: 0,
+          weight: 16
+        },
+        type: 'HEADERS', // by spec SYN_REPLY
+        writable: true
+      } , done);
+    })
```
Comment (indutny) [inline test/spdy/parser-test.js]: semicolon

**Comment by indutny** [comment]
Looks pretty good except some style nits! Thank you!

**Comment by diasdavid** [comment]
Awesome :) fixed all of the things you mentioned :)

**Comment by indutny** [comment]
Landed in 4eb339b, thank you so much!

**Comment by diasdavid** [comment]
wooo :D Awesome, thank you!
